internal/event/util.go

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package event
+
+import (
+	"reflect"
+	"strconv"
+)
+
+// MakeSubscriptionID creates subscription ID from a function
+func MakeSubscriptionID(fun interface{}) int64 {
+	value := reflect.ValueOf(fun)
+	if value.Kind() != reflect.Func {
+		panic("not a func")
+	}
+	return int64(value.Pointer())
```

Comment:
Godoc for [`Value.Pointer()`](https://golang.org/pkg/reflect/#Value.Pointer) says the following:
```
If v's Kind is Func, the returned pointer is an underlying code pointer, but not necessarily enough to identify a single function uniquely. The only guarantee is that the result is zero if and only if v is a nil func Value.
```

Considering this, relying on function pointer's address for subscription ids may be a bad idea. Not speaking of potential collisions with ids assigned to user subscriptions. WDYT?

Reply:
This function is only used internally. Note that there's a problem only with functions attached to structs, since that is just syntactic sugar that Go provides:
```go
type Foo struct {
}
func (*f Foo) Bar() bool {
 return false
}

foo1 := &Foo{}
foo2 := &Foo{}
// reflect.ValueOf(foo1.Bar).Pointer() == reflect.ValueOf(foo2.Bar)
```

Unless we use the the same struct function as a handler ID more than once in the same client, there's no problem.

Reply:
User event dispatcher is separate from the internal event dispatcher, so there's no subscription ID collision between them.

Reply:
> User event dispatcher is separate from the internal event dispatcher, so there's no subscription ID collision between them.

That's not obvious from the code and nothing enforces this invariant. This may become a source of subtle bugs in the future. I'd better go with a registry of constants for internal listeners and additional validation of absence of the listener id during the registration.

Reply:
Refactored listener stuff for DSSs: https://github.com/yuce/hazelcast-go-client/commit/15136dbdd26d8f9e1acdf9eba6c5ef62be6392d4